packages/zoe/src/contractSupport/zoeHelpers.js

```diff
@@ -107,14 +123,55 @@ export const makeZoeHelpers = (zcf) => {
       const rightAllocation = zcf.getCurrentAllocation(rightOfferHandle);
       const satisfied = (want, availableForTrade) =>
         getKeywords(want).every(keyword => {
-          const amountMath = helpers.getAmountMath(keyword);
+          const amountMath = zcf.getAmountMath(want[keyword].brand);
           return amountMath.isGTE(availableForTrade[keyword], want[keyword]);
         });
       return (
         satisfied(left.want, rightAllocation) &&
         satisfied(right.want, leftAllocation)
       );
     },
+
+    /**
+     * Compare two proposals for compatibility, mapping keywords. This returns
+     * true if the left offer would accept what the right offer is offering,
+     * presuming that the goods at each keyword are transferred across.
+     *
+     * @param {OfferHandle} leftOfferHandle
+     * @param {OfferHandle} rightOfferHandle
+     * @param {[Keyword]} keywords
+     * @returns boolean
+     */
+    canTradeWithMapKeywords: (leftOfferHandle, rightOfferHandle, keywords) => {
```

Comment:
This trade helper https://github.com/Agoric/agoric-sdk/pull/1147. I think the `trade` helper is the general form of swap, and if we change it to accommodate facet parameters, we can potentially get rid of swap.

Reply:
That seems pretty plausible. In addition to the usages you found, the barter contract could probably use something similar. 

I think we'd want to add mapped keywords to `trade`.  It's possible that `swap` is a degenerate form that doesn't specify keywords. 

It also seems like `canTradeWith()` (with or without keywords) would survive since sometimes the contracts want to ask rather than presuming. Though autoswap asks in order to know which variant to call. Perhaps it could presume if it could map keywords appropriately.

Are we far enough along that it makes sense to try to integrate `trade` in? How do you want to proceed?